// containers/disk/index.js
import React, {Component} from 'react';
import {connect} from 'react-redux';

import {fetchDiskById, addDiskToBasket} from 'actions';
import {getDiskById} from 'selectors';
import R from 'ramda';
import {Link} from 'react-router';
import BasketCart from 'components/basketCart';

class Disk extends Component {
  componentDidMount () {
    this.props.fetchDiskById(this.props.params.id);
  }

  renderFields () {
    const {disk} = this.props;
    const columnFields = R.compose(
      R.toPairs,
      R.pick([
        'name',
        'year',
        'price'
      ])
    )(disk)

    return columnFields.map(([key,value]) => (
      <div className='column' key={key}>
        <div className='ab-details-title'>
          <p>{key}</p>
        </div>
        <div className='ab-details-info'>
          {value}
        </div>
      </div>
    ))
  }

  renderContent () {
    const {disk} = this.props;
    return (
      <div className='thumbnail'>
        <div className='row'>
          <div className='col-md-6'>
              {this.renderFields()}
          </div>
          <div className='col-md-6'>
            <img 
              className='img-thumbnail' 
              src={disk.image} 
              alt={disk.name}  
            />
          </div>
        </div>
        <div className='caption-full'>
          <h4 className='pull-right'>${disk.price}</h4>
          <h4>{disk.name}</h4>
          <p>{disk.description}</p>
        </div>
      </div>
    )
  }

  renderSidebar () {
    const {disk, addDiskToBasket} = this.props;
    return (
      <div>
        <p className='lead'>Quick shop</p>
        <BasketCart />
        <div className='form-group'>
          <h1>{disk.name}</h1>
          <h2>${disk.price}</h2>
        </div>
        <Link
          to='/'
          className='btn btn-info btn-block'
        >
          Back to store
        </Link>
        <button
          type='button'
          className='btn btn-success btn-block'
          onClick={()=>addDiskToBasket(disk.id)}
        >
          Add to cart
        </button>
      </div>
    )
  }

  render () {
    const {disk} = this.props; 
    return (
      <div className='view-container'>
        <div className='container'>
          <div className='row'>
            <div className='col-md-9'>
              {disk && this.renderContent()}
            </div>
            <div className='col-md-3'>
              {disk && this.renderSidebar()}
            </div>
          </div>
        </div>
      </div>
    )
  };
}

const mapStateToProps = state => ({
  disk: getDiskById(state, state.diskPage.id)
})

const mapDispatchToProps = {
  fetchDiskById,
  addDiskToBasket
}

export default connect(mapStateToProps, mapDispatchToProps)(Disk);
